### ISSUE TITLE
Details about the usage of 3D labels?

### ISSUE THREAD

**Issue by revisitq** [pmj110119/RenderOcc#2]
The paper reported mIoU of the model supervised by 2D labels (depth map and semantic map) and 3D labels (occupancy semantic GT) is higher than only use 2D labels (26.11 v.s. 23.93). Could you tell me something details about use 2D&3D labels. In your model design, whether a voxel is occupied is determined by whether it's destiny bigger than a threshold. The semantic label of occupied voxel is derived from the semantic filed. In the case that only use 2D labels, you use semantic map to supervise semantic filed and use depth map to supervise density filed, the corresponding loss function is cross entropy and SILog. 
What I concern about is how to use 3D lables. In Nuscenes Datasets, the occupancy labels describe whether is a voxel occupied and it's semantic  (if occupied). For occupied voxel, it easy to supervise it by use cross entropy loss just like using 2D labels. But for unoccupied voxel, I don't have any idea about how to supervise it because the occupancy is determined by density. Could you tell me how to supervise unoccupied voxel with 3D labels in your design. Thanks!

**Comment by pmj110119** [comment]
When using 3D labels, **density** supervision can be accomplished in a binary classification manner. Here are two possible approaches for your consideration:

(1) Normalize the **density** (e.g. sigmoid) and supervise it with BCE loss.
(2) Use an MLP to output two channels representing the "occupied" and "empty" categories, and supervise it with softmax + CE loss. Utilize the first channel as the **density** (the second channel is discarded during inference).

Specifically, we used the second approach. Look forward to your further exploration!

**Comment by revisitq** [comment]
Got it! Thanks for your reply.

**Comment by revisitq** [comment]
> When using 3D labels, **density** supervision can be accomplished in a binary classification manner. Here are two possible approaches for your consideration:
> 
> (1) Normalize the **density** (e.g. sigmoid) and supervise it with BCE loss. (2) Use an MLP to output two channels representing the "occupied" and "empty" categories, and supervise it with softmax + CE loss. Utilize the first channel as the **density** (the second channel is discarded during inference).
> 
> Specifically, we used the second approach. Look forward to your further exploration!

More specifically, did you use `softplus`  after using softmax get density?

**Comment by pmj110119** [comment]
> > When using 3D labels, **density** supervision can be accomplished in a binary classification manner. Here are two possible approaches for your consideration:
> > (1) Normalize the **density** (e.g. sigmoid) and supervise it with BCE loss. (2) Use an MLP to output two channels representing the "occupied" and "empty" categories, and supervise it with softmax + CE loss. Utilize the first channel as the **density** (the second channel is discarded during inference).
> > Specifically, we used the second approach. Look forward to your further exploration!
> 
> More specifically, did you use `softplus` after using softmax get density?

Softmax is only used in the calculation of loss (torch.nn.CrossEntropyLoss). **Density** takes the value before 'softmax' and is not affected by it.

**Comment by revisitq** [comment]
Thanks! I got it.
